scene/gui/control.cpp

```diff
@@ -2202,62 +2202,60 @@ bool Control::is_clipping_contents() {
 
 // Theming.
 
-void Control::_propagate_theme_changed(Node *p_at, Control *p_owner, Window *p_owner_window, bool p_assign) {
+void Control::_propagate_theme_changed(Node *p_at, Control *p_owner, Window *p_owner_window, bool p_notify, bool p_assign) {
 	Control *c = Object::cast_to<Control>(p_at);
-
-	if (c && c != p_owner && c->data.theme.is_valid()) { // has a theme, this can't be propagated
-		return;
-	}
-
 	Window *w = c == nullptr ? Object::cast_to<Window>(p_at) : nullptr;
 
-	if (w && w != p_owner_window && w->theme.is_valid()) { // has a theme, this can't be propagated
+	if (!c && !w) {
 		return;
 	}
 
-	for (int i = 0; i < p_at->get_child_count(); i++) {
-		CanvasItem *child = Object::cast_to<CanvasItem>(p_at->get_child(i));
-		if (child) {
-			_propagate_theme_changed(child, p_owner, p_owner_window, p_assign);
-		} else {
-			Window *window = Object::cast_to<Window>(p_at->get_child(i));
-			if (window) {
-				_propagate_theme_changed(window, p_owner, p_owner_window, p_assign);
-			}
+	if (c) {
+		if (c != p_owner && c->data.theme.is_valid()) {
+			// Has a theme, so we don't want to change the theme owner,
+			// but we still want to propagate in case this child has theme items
+			// it inherits from the theme this node uses.
+			// See https://github.com/godotengine/godot/issues/62844.
+			p_assign = false;
 		}
-	}
 
-	if (c) {
 		if (p_assign) {
 			c->data.theme_owner = p_owner;
 			c->data.theme_owner_window = p_owner_window;
 		}
-		c->notification(Control::NOTIFICATION_THEME_CHANGED);
-		c->emit_signal(SceneStringNames::get_singleton()->theme_changed);
-	}
 
-	if (w) {
+		if (p_notify) {
+			c->notification(NOTIFICATION_THEME_CHANGED);
+		}
+	} else if (w) {
+		if (w != p_owner_window && w->theme.is_valid()) {
+			// Same as above.
+			p_assign = false;
+		}
+
 		if (p_assign) {
 			w->theme_owner = p_owner;
 			w->theme_owner_window = p_owner_window;
 		}
-		w->notification(Window::NOTIFICATION_THEME_CHANGED);
-		w->emit_signal(SceneStringNames::get_singleton()->theme_changed);
+
+		if (p_notify) {
+			w->notification(NOTIFICATION_THEME_CHANGED);
+		}
 	}
-}
 
-void Control::_theme_changed() {
-	_propagate_theme_changed(this, this, nullptr, false);
+	for (int i = 0; i < p_at->get_child_count(); i++) {
+		_propagate_theme_changed(p_at->get_child(i), p_owner, p_owner_window, p_notify, p_assign);
+	}
```

Comment:
I just put it here as I thought it looked nicer 🙂
Previously it was children first, then their ancestors, now the method's logic is applied to the ancestors first and then their children. Since there's no theme owner lookup or anything like that, it should matter.